pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -358,12 +366,35 @@ func (c *Controller) AppendWorkloadHandler(f func(*model.WorkloadInstance, model
 	return nil
 }
 
-// GetIstioServiceAccounts implements model.ServiceAccounts operation
+// GetIstioServiceAccounts implements model.ServiceAccounts operation.
+// The returned list contains all SPIFFE based identities that backs the service. For example,
+// - { "spiffe://cluster.local/bar@iam.gserviceaccount.com"}; when annotation is used on corresponding workloads.
+// - { "spiffe://cluster.local/ns/default/sa/foo" }; normal kubernetes cases
+// - { "spiffe://cluster.local/ns/default/sa/foo", "spiffe://trust-domain-alias/ns/default/sa/foo" };
+//   if the trust domain alias is configured.
 func (c *Controller) GetIstioServiceAccounts(svc *model.Service, ports []int) []string {
+	out := map[string]struct{}{}
 	for _, r := range c.GetRegistries() {
-		if svcAccounts := r.GetIstioServiceAccounts(svc, ports); svcAccounts != nil {
-			return svcAccounts
+		svcAccounts := r.GetIstioServiceAccounts(svc, ports)
+		for _, sa := range svcAccounts {
+			out[sa] = struct{}{}
 		}
 	}
-	return nil
+	result := []string{}
+	for k := range out {
+		result = append(result, k)
+	}
+	tds := []string{}
+	if c.meshHolder != nil {
+		mesh := c.meshHolder.Mesh()
+		if mesh != nil {
+			tds = mesh.TrustDomainAliases
+		}
+	}
+	expanded := spiffe.ExpandWithTrustDomains(result, tds)
```

Comment:
is aggregate controller the right place to expand trust domain alias? we may get unexpected results if we don't use aggregate and just directly use a registry.

I am not strongly against it, but lets at least document it as a comment maybe. And hopefully we will have tests that would catch this if someone breaks something around this in the future

Reply:
done. documented. we need e2e test anyway for sure . will do a in a separate pr.


currently we do init using aggregate controller from the manual testing as well as code reading.